compiler/src/parsing/parser.mly

```diff
@@ -389,6 +392,10 @@ app_expr:
 rcaret_rcaret_op:
   | lnonempty_list(RCARET) RCARET { (String.init (1 + List.length $1) (fun _ -> '>')) }
 
+construct_expr:
+  | type_id lparen lseparated_list(comma, expr) comma? rparen { Exp.construct ~loc:(to_loc $loc) $1 $3 }
+  | type_id %prec LPAREN { Exp.construct ~loc:(to_loc $loc) $1 [] }
```

Comment:
what's the difference between `lparen` and `LPAREN`?

Reply:
`LPAREN` is the literal token `(`, and `lparen` is the non-terminal rule we defined to handle them. It's got some extra rules around whitespace.

Reply:
But in this context we're specifying that the reduction should happen with the same precedence as encountering an LPAREN token, in this case instead shifting the LPAREN to try to parse a constructor with arguments instead of reducing a constant constructor.

Reply:
So tl;dr when the parser reads `Some` and the next token is `(`, keep parsing a constructor instead of immediately saying the entire constructor is just `Some`.